### PR TITLE
Improve switch room rendering

### PR DESCRIPTION
**Pull request by gsouquet**
Improves vector-im/element-web#4997

This reduces the time it takes to switch rooms without undergoing massive changes in `TimelinePanel` or `ScrollPanel`. There are a lot of room for improvement there but the changes required are a lot more invasive

Two main optimisations are applied here 
- Reducing the amount of elements in the DOM where possible
- Batching `setState` calls together and leveraging cached values to avoid a two step render where possible

When using a development build of react I managed to reduce the initial render from 570ms to 320ms which accounts for a 56% decrease in rendering time

### CODE REVIEW

File: src/components/structures/ContextMenu.tsx [hunk 1]
```diff
@@ -389,7 +389,7 @@ export class ContextMenu extends React.PureComponent<IProps, IState> {
     }
 
     render(): React.ReactChild {
-        return ReactDOM.createPortal(this.renderMenu(), getOrCreateContainer());
+        return ReactDOM.createPortal(this.renderMenu(), document.body);
```
Comment (turt2live) [inline src/components/structures/ContextMenu.tsx:392]: we mount/unmount using `getOrCreateContainer` elsewhere in the context menu code - what is this change accomplishing?

File: src/components/structures/RoomView.tsx [hunk 2]
```diff
@@ -83,6 +83,7 @@ import { objectHasDiff } from "../../utils/objects";
 import SpaceRoomView from "./SpaceRoomView";
 import { IOpts } from "../../createRoom";
 import {replaceableComponent} from "../../utils/replaceableComponent";
+import _ from 'lodash';
```
Comment (turt2live) [inline src/components/structures/RoomView.tsx:86]: we prefer the `import { fn }` syntax in hopes that treeshaking does the right thing (even though it doesn't).

File: src/components/structures/RoomView.tsx [hunk 3]
```diff
@@ -528,7 +530,18 @@ export default class RoomView extends React.Component<IProps, IState> {
     }
 
     shouldComponentUpdate(nextProps, nextState) {
-        return (objectHasDiff(this.props, nextProps) || objectHasDiff(this.state, nextState));
+        const hasPropsDiff = objectHasDiff(this.props, nextProps);
+
+        const newUpgradeRecommendation = nextState.upgradeRecommendation || {}
+
+        const state = _.omit(this.state, ['upgradeRecommendation']);
```
Comment (turt2live) [inline src/components/structures/RoomView.tsx:537]: fwiw, we have our own utility `objectExcluding` to accomplish this.

File: src/components/structures/RoomView.tsx [hunk 4]
```diff
@@ -528,7 +530,18 @@ export default class RoomView extends React.Component<IProps, IState> {
     }
 
     shouldComponentUpdate(nextProps, nextState) {
-        return (objectHasDiff(this.props, nextProps) || objectHasDiff(this.state, nextState));
+        const hasPropsDiff = objectHasDiff(this.props, nextProps);
+
+        const newUpgradeRecommendation = nextState.upgradeRecommendation || {}
```
Comment (turt2live) [inline src/components/structures/RoomView.tsx:535]: A comment for why we're explicitly pulling out this variable would be appreciated

File: src/components/structures/RoomView.tsx [hunk 5]
```diff
@@ -528,7 +530,18 @@ export default class RoomView extends React.Component<IProps, IState> {
...
+
+        const state = _.omit(this.state, ['upgradeRecommendation']);
+        const newState = _.omit(nextState, ['upgradeRecommendation'])
+
+        const hasStateDiff =
+            objectHasDiff(state, newState) ||
+            (newUpgradeRecommendation && newUpgradeRecommendation.needsUpgrade === true)
```
Comment (turt2live) [inline src/components/structures/RoomView.tsx:542]: it can't be null given the `or` above.

File: src/components/views/rooms/EventTile.tsx [hunk 6]
```diff
@@ -960,7 +979,7 @@ export default class EventTile extends React.Component<IProps, IState> {
             onFocusChange={this.onActionBarFocusChange}
         /> : undefined;
 
-        const timestamp = this.props.mxEvent.getTs() ?
+        const timestamp = this.props.mxEvent.getTs() && this.state.hover ?
```
Comment (turt2live) [inline src/components/views/rooms/EventTile.tsx:982]: we have a feature to always show timestamps that I think this would break.

**Review by turt2live — COMMENTED** [review]
overall looks fine - just a few relatively minor concerns. Thanks for poking this!

File: src/components/structures/ContextMenu.tsx (hunk 1 above)
Comment (gsouquet) [reply src/components/structures/ContextMenu.tsx:392]: It is pretty much to reduce the use extra nodes where they are not needed. Most of the time that container is not really required

File: src/components/structures/RoomView.tsx (hunk 3 above)
Comment (gsouquet) [reply src/components/structures/RoomView.tsx:537]: Shall we use the `lodash` implementation instead? It looks like we already go as far as loading the library in the codebase

Their implementation is thoroughly tested and has generally pretty good performance? 
I don't really mind either way

Comment (turt2live) [reply src/components/structures/RoomView.tsx]: I'm not personally opinionated on it: just wanted to mention that we have a thing that does this.

File: src/components/structures/ContextMenu.tsx (hunk 1 above)
Comment (turt2live) [reply src/components/structures/ContextMenu.tsx:392]: not sure I fully understand: we are still referencing `getOrCreateContainer()` within the class and outside of `render()`: is this supposed to be fixed too?

**Review by turt2live — COMMENTED** [review]
github wants words in this box

**Comment by turt2live** [comment]
also merge conflicts (already) :(

**Review by turt2live — APPROVED** [review]

**Pull request merged**